Cython/Compiler/ModuleNode.py

```diff
@@ -1592,13 +1593,23 @@ def generate_new_function(self, scope, code, cclass_entry):
             code.putln("new((void*)&(p->%s)) %s();" % (
                 entry.cname, decl_code))
 
+        if scope.internal_mode == 2 and py_attrs:
+            # create a hybrid "optional initialization" where the kwds
+            # argument is used to signal that initialization happens.
+            # This is needed for pickleable closures, where the attributes
+            # need to be uninitialized when created in functions (to match
+            # existing behaviour) but initialized when unpickling (to
+            # avoid segfaults decrefing NULL)
+            code.putln("if (k) {")
         for entry in py_attrs:
             if entry.name == "__dict__":
                 needs_error_cleanup = True
                 code.put("p->%s = PyDict_New(); if (unlikely(!p->%s)) goto bad;" % (
                     entry.cname, entry.cname))
             else:
                 code.put_init_var_to_py_none(entry, "p->%s", nanny=False)
+        if scope.internal_mode == 2 and py_attrs:
+            code.putln("}")
```

Comment:
I think it would be better to use enum (or at least strings) to indicate the state

Reply:
enum requires a little more text and imports but is less error-prone

Oh yeah .. there is no enum in python 2.7 ... but we can use the enum34 backport!

Reply:
> but we can use the enum34 backport!

No. I think we try quite hard not to require extra dependencies. So while I like enums it isn't worth adding a dependency for that.

Reply:
> I think we try quite hard not to require extra dependencies.

Is there any reason for this? If our code is cleaner and limits unnecessary freedom, then the quality of the product itself will improve.

Not to mention the fact that we can probably refuse to support running cython itself on python 2.7, while still allowing compilation for this version.

Reply:
This PR isn't really the place to debate this so I'll reply once and then leave the topic, but:

> Is there any reason for this? If our code is cleaner and limits unnecessary freedom, then the quality of the product itself will improve.

Adding dependencies would Cython harder to install (you can no longer tell users "just install Cython"...), you have to track versions for dependencies (have they changed and broken anything, is it a problem if users have another a different version installed?), and you can't necessarily control the dependencies they bring in. Of course it's a trade-off and there's definitely advantages to using good code written elsewhere, but for something like Cython that's a dependency of a lot of other stuff there's a big advantage to having very minimal decisions. (Of course, this wasn't my decision when it was originally made, and I can't know all the motivations of the people who made it...)

> Not to mention the fact that we can probably refuse to support running cython itself on python 2.7, while still allowing compilation for this version.

That's probably as good as dropping support entirely: if I want a Cython module compiling with a specific version of Python I call `pythonx.x setup.py ...`, or `cythonize-x.x` (which uses that version of Python), or use a Jupyter notebook with that version of Python. Of course it's possible to do the steps separately but it's hard and most people don't.

Reply:
> This PR isn't really the place to debate this

True, but where is that place?

> Of course it's possible to do the steps separately but it's hard and most people don't.

I think it's the matter of giving the users such mechanism.
Like with the `ast.parse` (`feature_version` argument), we can just apply different rules and restrictions depending on the versions for what we compile.
I don't know how long it will take to end support for python 2, it seems to me long enough to create an artificial support mechanism for compiling for earlier versions.

There may be some caveats with compiling for python 2 with the library of python 3, but we can either request the presence of the python 2 files, or (the better but longer/harder way) just simulate or copypaste some implementation details, that are not present in python 3

Reply:
> > This PR isn't really the place to debate this
> 
> True, but where is that place?

The cython-devel mailing list. But please don't ask for this specific addition as I would only reject it.

> Not to mention the fact that we can probably refuse to support running cython itself on python 2.7, while still allowing compilation for this version.

This would make Cython substantially more difficult to install and test. Also not up for discussion right now. Py2.7 support will be removed in Cython 3.1, so there is no need to even think about this.

Reply:
> But please don't ask for this specific addition as I would only reject it.

I don't consider this to be a reason, please, don't think about me like that.